singularity-eos/eos/eos_polynomial.hpp

```diff
@@ -0,0 +1,212 @@
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return 0.0;
+  }
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    const Real mu = rho/_rho0 - 1;
+    if (mu >= 0) // Compression
+      return  (1+mu)*(
+                              _a1 + 2*_a2c*mu + 3*_a3*mu*mu +
+                      sie * ( _b1 + 2*_b2c*mu + 3*_b3*mu*mu )
+                      );
+    else
+      return  (1+mu)*(
+                              _a1 + 2*_a2e*mu + 3*_a3*mu*mu +
+                      sie * ( _b1 + 2*_b2e*mu + 3*_b3*mu*mu )
+                      );
+  }
+
+  // gss: what does this do?
+  PORTABLE_INLINE_FUNCTION void FillEos(Real &rho, Real &temp, Real &energy, Real &press,
+                                        Real &cv, Real &bmod, const unsigned long output,
+                                        Real *lambda = nullptr) const;
+
+
+  // gss: what does this do?
+  PORTABLE_INLINE_FUNCTION
+  void ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                              Real &bmod, Real &dpde, Real &dvdt,
+                              Real *lambda = nullptr) const {
+    // use STP: 1 atmosphere, room temperature
+    rho = _rho0;
+    temp = 0.0;
+    sie = 0.0;
+    press = 0.0;
+    cv = 0.0;
+    bmod = 0.0;
+    dpde = 0.0;
+    dvdt = 0.0;
+  }
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(Polynomial)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    return 0;
+  }
+  static inline unsigned long max_scratch_size(unsigned int nelements) { return 0; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    printf("Polynomial EOS Parameters:\na0 = %g\na1    = %g\n", _a0, _a1);
+  }
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const {
+    sie = 0.0;
+    rho = 0.0;
+  }
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Polynomial"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  Real _a0, _a1, _a2c, _a2e, _a3, _b0, _b1, _b2c, _b2e, _b3;
+  // reference values
+  Real _rho0;
+  // gss: Dont think i need these
+  //  static constexpr const Real _T0 = ROOM_TEMPERATURE;
+  //  static constexpr const Real _P0 = ATMOSPHERIC_PRESSURE;
+  // static constexpr const char _eos_type[] = {"IdealGas"};
+
+  // gss: what does this do?
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::specific_internal_energy;
```

Comment:
It exposes some introspection to host codes. They can ask whether it's more performant to pass in energy or temperature as an independent variable... i.e., if you have rho-T tables, then it's more performant to pass in temperature.